cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
+	// Get the CoreDNS ConfigMap. If not present, Create a new default ConfigMap.
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Get(kubeadmconstants.CoreDNSConfigMap, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil
+		}
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Create(cm); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
+				return errors.Wrap(err, "unable to create configmap")
+			}
+		}
+	}
+	// Get the Corefile and installed CoreDNS version.
+	corefile, currentInstalledCoreDNSVersion, err := GetCoreDNSInfo(client)
+	if err != nil {
+		return errors.Wrap(err, "unable to fetch CoreDNS current installed version and ConfigMap.")
+	}
+	// Check if the current Configuration present is the default (unmodified) version.
+	// If not, try and migrate the Config.
+	if !migration.Default("", corefile) {
+		updatedCorefile, err := migration.Migrate(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile, false)
+		if err != nil {
+			return errors.Wrap(err, "unable to migrate CoreDNS Configmap")
+		}
+
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(&v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      kubeadmconstants.CoreDNSConfigMap,
+				Namespace: metav1.NamespaceSystem,
+			},
+			Data: map[string]string{
+				"Corefile":          updatedCorefile,
+				"Corefile-previous": corefile,
+			},
+		}); err != nil {
+			return errors.Wrap(err, "unable to update configmap")
+		}
+		changes, err := migration.Deprecated(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile)
+		if err != nil {
+			return errors.Wrap(err, "unable to get list of notifications")
+		}
+		klog.Infof("the CoreDNS Configuration has been migrated and applied: %v.", updatedCorefile)
+		klog.Infof("the old migration has been saved in the CoreDNS ConfigMap under the name [Corefile-previous]")
+		klog.Infof("The changes in the new CoreDNS Configuration are as follows:")
+		for _, change := range changes {
+			klog.Infof("%v", change.ToString())
+		}
+	}
+	// If the current Configuration present is the default (unmodified) version, just update it to the latest default.
+	if migration.Default("", corefile) {
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(cm); err != nil {
+			return errors.Wrap(err, "unable to update configmap")
+		}
+	}
+	return nil
+}
+
+// GetCoreDNSInfo gets the current CoreDNS installed and the current Corefile Configuration of CoreDNS.
+func GetCoreDNSInfo(client clientset.Interface) (string, string, error) {
+	coreDNSConfigMap, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(kubeadmconstants.CoreDNSConfigMap, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return "", "", err
+	}
+	corefile := coreDNSConfigMap.Data["Corefile"]
+
+	dnsType, currentCoreDNSversion, err := DeployedDNSAddon(client)
```

Comment:
i think it is possible to avoid checking if the addon type is CoreDNS in `GetCoreDNSInfo`.
the function should be called only if the addon is known to be CoreDNS.